### Scripts used for TWAS

#### 1. Association Study with Individual-level GWAS Data
- Asso_Study_01.py

#### 2. Association Study with Summary-level GWAS Data
- summary_stat.py, Asso_Study_02.sh

#### 3. Generate reference genotype covariance file
- covar_calculation.py, covar_calculation.sh
- Input file of genome block annotation is required for generating reference genotype covaraince file. Example block annotation file for European samples is provided `./TIGAR/example_data/block_annotation_EUR.txt`. 
- The block annotation file is a tab delimited text file with head row of `CHROM Start End File`, denoting the chromosome number, starting position, ending position, and corresponding reference VCF file name under specified `--geno_path`. Reference VCF files shall be of one per chromosome, or one for the whole genome-wide variants. 
- Block annotation files of other ethnicities can be adopted from the genome segmentation generated by `LDetect`, https://bitbucket.org/nygcresearch/ldetect-data/src/master/.

